Fail when Should().[Not]Throw used on async void Action

zarenner opened the pull request:
## IMPORTANT 
SINCE FLUENT ASSERTIONS IS CURRENTLY UNDER HEAVY REFACTORING FOR [VERSION 5.0](https://github.com/fluentassertions/fluentassertions/issues/463), WE CURRENTLY ONLY ACCEPT BUGFIX REQUESTS. SORRY FOR THE INCONVENIENCE

* [ ] The [Pull Request](https://help.github.com/articles/using-pull-requests) is targeted at the `master` branch.
* [ ] The code complies with the [Coding Guidelines for C# 3.0, 4.0 and 5.0](http://www.csharpcodingguidelines.com/)/.
* [ ] The changes are covered by a new or existing set of unit tests which follow the Arrange-Act-Assert syntax such as is used [in this example](https://github.com/fluentassertions/fluentassertions/blob/daaf35b9b59b622c96d0c034e8972a020b2bee55/Tests/FluentAssertions.Shared.Specs/BasicEquivalencySpecs.cs#L33).
* [ ] If the contribution affects the documentation, please include your changes to [**documentation.md**](https://github.com/fluentassertions/fluentassertions/blob/master/docs/documentation.md) in this pull request so the documentation will appear on the [website](http://fluentassertions.com/documentation.html).

#711

zarenner commented on Src/FluentAssertions/Specialized/ActionAssertions.cs:
```diff
@@ -119,6 +128,14 @@ private Exception InvokeSubjectWithInterception()
+        {
+            if (Subject.GetMethodInfo().HasAttribute<AsyncStateMachineAttribute>())
+            {
+                throw new InvalidOperationException($"Cannot use {assertionMethodName} on an Action representing an async void method.");
```
Let me know if you prefer a more descriptive message indicating that Action very likely should have been Func<Task>

dennisdoomen replied:
Yes, I think emphasizing the possible solution does help the user understanding his mistake.

zarenner commented on Src/FluentAssertions/Specialized/ActionAssertions.cs:
```diff
@@ -35,6 +38,8 @@ public ActionAssertions(Action subject, IExtractExceptions extractor)
         public ExceptionAssertions<TException> Throw<TException>(string because = "", params object[] becauseArgs)
             where TException : Exception
         {
+            FailIfSubjectAsyncVoid("Throw");
```
Let me know if you disagree with adding the check on Should().Throw() as well. It doesn't need it as much as NotThrow IMO, but could help users avoid confusion while writing tests.

Meir017 replied:
Maybe use `nameof`

zarenner replied:
afaik nameof doesn't let you capture the correct (generic or non-generic) NotThrow method. Not that I expect one name to change and not the other, but still seemed wrong.

zarenner commented on Src/FluentAssertions/Specialized/ActionAssertions.cs:
```diff
@@ -119,6 +128,14 @@ private Exception InvokeSubjectWithInterception()
 
+        private void FailIfSubjectAsyncVoid(string assertionMethodName)
+        {
+            if (Subject.GetMethodInfo().HasAttribute<AsyncStateMachineAttribute>())
```
Not sure what FA's stance on reflection is - this is the only way I'm aware of to do this check though.

jnyrup commented:
If `Throw` gets it, please also add it to`ThrowExactly`, otherwise the exception message would mention `Throw` when calling `ThrowExactly`.

Shouldn't `FailIfSubjectAsyncVoid` be `FailIfSubjectIsAsyncVoid`?

zarenner commented:
Updated exception message to use term "action assertions" and suggest fix. Renamed FailIfSubjectAsyncVoid to FailIfSubjectIsAsyncVoid. Added test for ThrowExactly.
